qiskit/quantum_info/operators/operator.py

```diff
@@ -514,7 +514,7 @@ def _init_instruction(cls, instruction):
     @classmethod
     def _instruction_to_matrix(cls, obj):
         """Return Operator for instruction if defined or None otherwise."""
-        if not isinstance(obj, Instruction):
+        if not isinstance(obj, Operation):
```

Comment:
Oh, this is a very good point. On the one hand, we definitely want to construct an `Operator` from a quantum circuit that contain a `Clifford` or some other higher-level object:
```
cliff = Clifford(...)
qc = QuantumCircuit(5)
qc.append(cliff, [0,1,2])
op = Operator(qc)
```
On the other hand, the current code only works because `Clifford` happens to have a `to_matrix()` method, which is not required for a general `Operation`. So my "fix" above would not carry-over to future higher-level objects. Thinking aloud, the main code (see https://github.com/Qiskit/qiskit-terra/blob/2bab09c1aae84e5bf38ba52fa3a272667c1887cc/qiskit/quantum_info/operators/operator.py#L529) first checks that an `Operation` has or can have a matrix, then if it's a `Barrier`, then if it has a `definition`, and then raises an error. So ideally (as higher-level objects do not need to carry a `definition` either) we should call the `HighLevelSynthesis` pass here (or more precisely whichever routine that uses to synthesize Cliffords). But this is ugly -- as `quantum_info/Operator` would depend on transpiling routines, leading to all sorts of cyclic dependencies (run-time import and pylint-disable would fix this, but nonetheless). And I don't see another way. @mtreinish , @kdk , what do you think about this?

Reply:
What about doing:

```suggestion
        if not isinstance(obj, Operation) and hasattr(obj, "to_matrix"):
```

that way we still work with operations like clifford that have a defined matrix. We probably should include this in the operation class docs too as an optional function if we start making this an implicit part of the class (I do worry about bloating operation and having it just repeat `Instruction` though).

Reply:
Thanks, I see now. It should be ok that we cannot construct an `Operator` for an arbitrary quantum circuit, and in particular for circuits containing some higher-level objects (and we already can't do it for circuits containing conditional gates). Though, I completely agree, we shouldn't break this functionality for Cliffords.

It's also no clear to me if constructing `Operators` for higher-level objects based on `HighLevelSynthesis` is correct (i.e. what I was suggesting earlier) -- what if the synthesis method uses ancillas?

So I am happy to adopt the above suggestion (though, hmm, it should be `not hasattr`...). 

Agreed, I would also like to avoid bloating `Operation`. So, if I understood your suggestion correctly, I will simply include a comment in the Operation class docs.